src/core/meshcop/joiner.cpp

```diff
@@ -0,0 +1,293 @@
+    static_cast<Joiner *>(aContext)->HandleUdpReceive(*static_cast<Message *>(aMessage),
+						      *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void Joiner::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    (void)aMessageInfo;
+
+    mNetif.GetDtls().Receive(aMessage, aMessage.GetOffset(), aMessage.GetLength() - aMessage.GetOffset());
+
+    if (mNetif.GetDtls().IsConnected())
+    {
+        SendJoinerFinalize();
+    }
+}
+
+void Joiner::HandleUdpTransmit(void *aContext)
+{ 
+    otLogInfoMeshCoP("Joiner::HandleUdpTransmit\r\n");
+    static_cast<Joiner *>(aContext)->HandleUdpTransmit();
+}
+
+void Joiner::HandleUdpTransmit(void)
+{
+    ThreadError error = kThreadError_None;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit(mTransmitMessage != NULL, error = kThreadError_NoBufs);
+
+    otLogInfoMeshCoP("transmit %d\r\n", mTransmitMessage->GetLength());
+
+    memset(&messageInfo, 0, sizeof(messageInfo));
+    messageInfo.GetPeerAddr().mFields.m16[0] = HostSwap16(0xfe80);
+    messageInfo.GetPeerAddr().SetIid(mJoinerRouter);
+    messageInfo.mPeerPort = 1000;
+    messageInfo.mInterfaceId = 1;
+
+    SuccessOrExit(error = mSocket.SendTo(*mTransmitMessage, messageInfo));
+
+exit:
+
+    if (error != kThreadError_None && mTransmitMessage != NULL)
+    {
+        Message::Free(*mTransmitMessage);
+    }
+
+    mTransmitMessage = NULL;
+}
+
+void Joiner::SendJoinerFinalize(void)
+{
+    Coap::Header header;
+    MeshCoP::StateTlv *stateTlv;
+    uint8_t buf[128];
+    uint8_t *cur = buf;
+
+    header.Init();
+    header.SetVersion(1);
+    header.SetType(Coap::Header::kTypeConfirmable);
+    header.SetCode(Coap::Header::kCodePost);
+    header.SetMessageId(0);
+    header.SetToken(NULL, 0);
+    header.AppendUriPathOptions(OPENTHREAD_URI_JOINER_FINALIZE);
+    header.AppendContentFormatOption(Coap::Header::kApplicationOctetStream);
+    header.Finalize();
+    memcpy(cur, header.GetBytes(), header.GetLength());
+    cur += header.GetLength();
+
+    stateTlv = reinterpret_cast<MeshCoP::StateTlv *>(cur);
+    stateTlv->Init();
+    stateTlv->SetState(MeshCoP::StateTlv::kAccept);
+    cur += sizeof(*stateTlv);
+
+    mNetif.GetDtls().Send(buf, static_cast<uint16_t>(cur - buf));
+
+    otLogInfoMeshCoP("Sent joiner finalize\r\n");
+}
+
+void Joiner::ReceiveJoinerFinalizeResponse(uint8_t *buf, uint16_t length)
+{
+    Message *message;
+    Coap::Header header;
+
+    VerifyOrExit((message = Message::New(Message::kTypeIp6, 0)) != NULL, ;);
+    SuccessOrExit(message->Append(buf, length));
+    SuccessOrExit(header.FromMessage(*message));
+
+    VerifyOrExit(header.GetType() == Coap::Header::kTypeAcknowledgment &&
+                 header.GetCode() == Coap::Header::kCodeChanged &&
+                 header.GetMessageId() == 0 &&
+                 header.GetTokenLength() == 0, ;);
+
+    otLogInfoMeshCoP("received joiner finalize response\r\n");
+
+    Close();
```

Comment:
@jwhui , I misunderstood the last Alert message meaning while reading the captured pcap file. Thanks for your help and education!

For section 8, they only verified till DTLS session close.
For section 9, they verify beginning from the topology has been formed.
To my understanding, we can keep "thread start" separately, but need to provide a way for the user to check when to begin the attach process.

This is the pcap file attached in case for your need.
You can check 3422 to 3675 for the DTLS handshake and Joiner protocol.
[thread_sniffer_fw_Sep1-2016.tar.gz](https://github.com/openthread/openthread/files/454328/thread_sniffer_fw_Sep1-2016.tar.gz)
We can see a lot of retransmissions there.